[SIP-5] QueryBuilder in the client for granularity and groupby in word cloud

xtinec opened the pull request:
- Lay the structure of the QueryContext builder in the client
- QueryContext builder composes different portions of the queryContext object (to be sent to server from the client) from the datasourceBuilder and the queryObjectBuilder.
- The datasourceBuilder builds the datasource id and type by parsing them from the datasource field in formdata
- The queryObjectBuilder currently only builds the groupby field. It will further compose the queryObject from sub query builders in future PRs.
- Create a buildQuery method for WordCloud and override the groupby value with the value of series from formdata.
@kristw @williaster @conglei @mistercrunch

Commit:
[SIP-5] QueryBuilder in the client for groupby field

- Lay the structure of the QueryContext builder in the client
- QueryContext builder composes different portions of the queryContext object (to be sent to server from the client) from the datasourceBuilder and the queryObjectBuilder.
- The datasourceBuilder builds the datasource id and type by parsing them from the datasource field in formdata
- The queryObjectBuilder currently only builds the groupby field. It will further compose the queryObject from sub query builders in future PRs.
- Create a buildQuery method for WordCloud and override the groupby value with the value of series from formdata.

williaster commented on superset/assets/src/query/formData.ts:
```diff
@@ -0,0 +1,7 @@
+// Type signature and utility functions for formData
+// It will be gradually filled out as we build out the query object
```
🙌

williaster commented on superset/assets/src/query/datasourceBuilder.ts:
```diff
@@ -0,0 +1,19 @@
+  type: DatasourceType;
+}
+
+export default function build(formData: FormData): Datasource {
```
could we match the name the file for easier debugging? I want to add the lint rule for this.

xtinec replied:
Yep! so `datasourceBuilder` as the function name? Do we care for a verb like `buildDatasource`? Preference? I don't have a strong opinion about this. 😄

kristw replied:
If it is a single function, I would prefer verb.

williaster commented on superset/assets/src/query/queryContextBuilder.spec.ts:
```diff
@@ -0,0 +1,24 @@
+  });
+
+  it('should call datasourceBuilder to build datasource', () => {
+    expect(datasourceBuilderSpy).toHaveBeenCalled();
```
nit -- possibly more robust if we specify the # times called with `toHaveBeenCalledTimes(1)` but not huge deal. I think the spies would have to be reset per-call tho.

kristw replied:
I have question about where to put test files. I like having them side-by-side but both `incubator-superset` and `superset-ui` put test files in separate folders. Perhaps this PR can follow the same structure to put tests under `test`?

xtinec replied:
Would we move everything under visualization to `superset-ui` eventually? If so, perhaps, we could put the query builder tests and the `buildQuery` tests for each visualization together in a dedicated directory (call it `superset-ui`?) under `spec/javascripts`? This way we could move them all together. Thoughts?

kristw replied:
sgtm. That's how I did it for the color modules.

williaster commented on superset/assets/src/query/queryContextBuilder.ts:
```diff
@@ -0,0 +1,15 @@
+import buildDatasource, { Datasource } from './datasourceBuilder';
```
should we call the files themselves `buildDatasource` and `buildQueries` if that's the most descriptive way to import them?

xtinec replied:
I like it! Then we have the default exports match the file names exactly. 👍

williaster commented on superset/assets/src/visualizations/wordcloud/buildQuery.ts:
```diff
@@ -0,0 +1,14 @@
+import buildQueryContext, { FormData, QueryContext } from 'src/query';
+
+export default function buildQuery(formData: FormData): QueryContext {
```
do you think it's worthwhile to specify specialized `FormData` interfaces per-vis so it's clear which (sub-set of) fields a given vis requires/depends on?

xtinec replied:
Good suggestion! 

Fiddling with this to see if we can compose the interface definitions (much like the query builders).

williaster reviewed (commented):
this is awesome! 🏆 cool to learn more `TypeScript`! 💁‍♀️

had a couple minor comments about naming + a question about per-vis `FormData` interfaces.

kristw reviewed (commented):
Excited about TypeScript too. 
What are the difference between `QueryObject` and `QueryContext` and their relationship?
Which one is sent to server? 
How will this change the diagram in
https://docs.google.com/document/d/1nKnBxKTOrHW1bqidOE-9cMPH7NZFAYDqLh8DAcVcGVA/edit#

xtinec commented:
`QueryContext` is sent to the server. It contains an array of `QueryObject`s and the corresponding `datasource` (with `id` and `type`). 
I think we can update the `query_object` box in the diagram to say `query_context (datasource + query_object[])`.

Commit:
Addressing PR comments
- Rename query builder files and their default exports
- Move tests into spec/javascripts/superset-ui for easy mass migration to superset-uiin the future
- Define viz specific formData and export formData for each viz type as intersection type of the generic formData and the viz specific one
- Specify the type signature for sqla and druid based data sources

kristw commented on superset/assets/src/query/buildDatasource.ts:
```diff
@@ -10,7 +10,7 @@ export interface Datasource {
 }
 
-export default function build(formData: FormData): Datasource {
+export default function buildDatasource(formData: FormData): Datasource {
```
- Instead of `buildDatasource` that returns interface `Datasource`, what do you think about making a class instead? then add `.toString()` which returned string concatenated form and `.toObject()` to return plain object.

- We may want to rename this `Datasource` to `DatasourceKey` because there will be another `Datasource` data structure that wraps returned value from `/superset/fetch_datasource_metadata?datasourceKey={key}`. That one contains full metadata about datasource and is used by chart. 

https://docs.google.com/presentation/d/1yJQWGr5p7BnfSoaMaM3HRB7EUKU067pkdHKnAwzzdXs/edit?usp=sharing

xtinec replied:
I typically lean towards interface as it is lighter-weight than class but I really like the idea of using a class for DataSource as we do have the need to consume a datasource as both a `datasourceKey` and a plain js object. 👍

xtinec replied:
Btw, still maintaining the interface definition as we need to type the query context payload going back to the server. So we could have a `DataSource` class and in it two methods, `toKey()` to convert a datasource for a datasourceKey `string` and `toObject()` to convert it to a plain js object typed with the interface.

kristw replied:
Great I like the revision with class. BTW, what do you think about `DatasourceKey` and `Datasource` issue I mentioned above?

The one you are working on here is what I refer to as `DatasourceKey`. It only contains enough information `{ id, type }` to identify the datasource in database.

There is another `Datasource` we will run into in a separate PR at some point, which is the full metadata of datasource: `{ verbose_map, columns, ... }`. Both of them are referred to in the code as `datasource` now, which is quite confusing.

xtinec replied:
Aha, I see. That is indeed very confusing. Totally missed the difference. hehe.

I could rename the class to `DatasourceKey` unless we see the two classes merging at some point.

kristw commented on superset/assets/src/query/buildQueryContext.ts:
```diff
@@ -1,13 +1,13 @@
-import buildDatasource, { Datasource } from './datasourceBuilder';
+import buildDatasource, { Datasource } from './buildDatasource';
+import buildQueries, { QueryObject } from './buildQueryObject';
```
Is it `buildQueries` or `buildQueryObject`?

xtinec replied:
`buildQueries` is probably better since we'll always return an array. I'll rename the file and the default export.

mistercrunch commented on superset/assets/src/query/buildDatasource.ts:
```diff
@@ -0,0 +1,19 @@
+  const [id, type] = formData.datasource.split('__');
+  return {
+    id: parseInt(id, 10),
+    type: type === 'table' ? DatasourceType.Table : DatasourceType.Druid,
```
Is this line useful? Also if someone in the wild has a custom connector it would break here.

Commit:
Addressing additional PR comments.
- Introduce a Datasource class and leverage Typescript's declaration merging of the interface
by the same name.
- Let Typescript infer the return type of various builders instead of specifying them explicitly

kristw reviewed (commented):
I have a broader question about `QueryContext` and vis-specific `buildQuery` usage.

There seems to be two ways we can go with this:

#### 1) On use, lookup and call vis-specific `buildQuery`, which calls the generic `buildQueryContext` under the hood. 

```js
import { getChartBuildQueryRegistry } from `@superset-ui/chart`;

function query(formData) {
  // formData.visType is 'big_number', for example.
  const buildQueryContext = getChartBuildQueryRegistry().get(formData.visType); 
  if (buildQueryContext) {
    const queryContext = buildQueryContext(formData);
    return SupersetClient.post('/api/v1/query', queryContext);
  } 
  return Promise.reject('Invalid visType');
}
```

#### 2) `buildQueryContext` encapsulates the logic for finding the vis-specific `buildQuery` under the hood. 

```js
import { buildQueryContext } from `@superset-ui/query`;

function query(formData) {
  const queryContext = buildQueryContext(formData);
  return queryContext 
    ? SupersetClient.post('/api/v1/query', queryContext)
    : Promise.reject('Invalid visType');
}
```

From the current code, we are going in 1) direction if I understanding correctly?

kristw commented on superset/assets/src/query/buildQueries.ts:
```diff
@@ -0,0 +1,18 @@
+// buildQuery method for each viz type (see `wordcloud/buildQuery.ts` for an example).
+// Note the type of the formData argument passed in here is the type of the formData for a
+// specific viz, which is a subtype of the generic formData shared among all viz types.
+export default function buildQueries<T extends FormData>(formData: T): QueryObject[] {
```
When would this return multiple query objects?

williaster replied:
^I think that's correct. E.g., time series with time shift is 2 queries on the backend now, so I think it will become 2 queries on the client.

xtinec replied:
Good question. This was also bothering me as I was writing this method up. It seems each viz-specific `buildQuery` would need to provide the query count to the generic one the # of common queries to construct.

Thinking restructuring this..

williaster replied:
ah good question. I guess for a time shift it would be plausible to auto-generate the second query, but still potentially tricky 🤔

kristw commented on superset/assets/src/visualizations/wordcloud/buildQuery.ts:
```diff
@@ -0,0 +1,15 @@
+import buildQueryContext from 'src/query';
+import { FormData } from './formData';
+
+export default function buildQuery(formData: FormData) {
```
I thought about this more as someone who needs to write plugin. Might be cleaner from plugin developer standpoint if developer only has to write

```js
export default function buildQuery(formData: FormData) {
  return buildQueryContext(formData, (baseQueryObject) => [{
     ...baseQueryObject,
     groupby: [formData.series],
  }]);
}
```

Assuming it will always be the same process for  `datasource` and the plugin should only touch the `queries`.

williaster replied:
Even this would become slightly more complex when modifying the `queries` tho, at that point they seem pretty similar to me?

```javascript
export default function buildQuery(formData: FormData) {
  return buildQueryContext(formData, (baseQueryObject) => {
    const { queries, ...rest } = baseQueryObject;
    const [ query ] = queries;

    return {
      ...rest,
      queries: [{
        ...query,
        groupby: [formData.series],
      }],
    };
  };
}
```

kristw replied:
Umm, I mean to abstract that complex part into `buildQueryContext`, so vis-specific `buildQuery` 
- does not need to touch `datasource`. 
- does not modify `queries` but receive `baseQueryObject` that is the single `queryObject` currently returned in `queryContext.queries` and use that to construct `queries` and return. 

Current `buildQueryContext`

```js
function buildQueryContext(formData: FormData): QueryContext {
  return {
    datasource: buildDatasource(formData),
    queries: buildQueries(formData),
  };
}
```

With my proposed change

```js
function buildQueryContext(formData: FormData, buildQueries: Function): QueryContext {
  return {
    datasource: buildDatasource(formData),
    queries: buildQueries(buildQueryObject(formData)),
  };
}
```

xtinec replied:
I like these guiding principles and the usage of higher-order function in the proposal:
> vis-specific buildQuery
> - does not need to touch datasource.
> - does not modify queries but receive baseQueryObject that is the single queryObject currently > in queryContext.queries and use that to construct queries and return.

Relating to the `BuildQueryRegistery` discussion, this approach would work nicely (we would register each viz-specific `buildQuery` function).

One small tweak I'm thinking is to define `BaseQueryObject` as a class (instead of a plain js object) so we could more easily add utility functions viz-specific `buildQuery` may need to enrich their queries. For example, imagine generating filter specific structures that can be accessed by viz-specific `buildQuery`s to construct their own pre-processing queries. We would have `.toObject()` in the class as well for accessing the plain js object.

xtinec replied:
Okay, I deciding against the small tweak as I'm writing it. I think we can add it once the need arises. Going to leave baseQueryObject as an interface for now.

kristw replied:
LGTM

xtinec commented on superset/assets/src/visualizations/wordcloud/formData.ts:
```diff
@@ -0,0 +1,10 @@
+
+// FormData for wordcloud contains both common properties of all form data
+// and properties specific to wordcloud vizzes
+export type FormData = GenericFormData & WordCloudFormData;
```
👍

xtinec commented:
Correct re the `QueryContext` and vis-specific `buildQuery` usage. We're going in the direction of approach 1. 

Approach 2 is also valid but it not going to be easy to figure out which `buildQuery` to use without giving each `formData` a `type` property indicating which viz they're intended for. And the `type` property is essentially the buildQueryRegistry.

Any opinion re approach 1 vs 2?

kristw commented:
See another thread in the PR that Chris and I discuss. That might be helpful in shaping the restructure.
On Nov 15, 2018, 08:32 -0800, Christine Chambers <notifications@github.com>, wrote:
> @xtinec commented on this pull request.
> In superset/assets/src/query/buildQueries.ts:
> > @@ -0,0 +1,18 @@
> +import { FormData, getGranularity } from './formData';
> +
> +// TODO: fill out the rest of the query object
> +export interface QueryObject {
> +  granularity: string;
> +  groupby?: string[];
> +}
> +
> +// Build the common segments of the query objects (e.g. the granularity field derived from
> +// either sql alchemy or druid). The segments specific to each viz type is constructed in the
> +// buildQuery method for each viz type (see `wordcloud/buildQuery.ts` for an example).
> +// Note the type of the formData argument passed in here is the type of the formData for a
> +// specific viz, which is a subtype of the generic formData shared among all viz types.
> +export default function buildQueries<T extends FormData>(formData: T): QueryObject[] {
> Good question. This was also bothering me as I was writing this method up. It seems each viz-specific buildQuery would need to provide the query count to the generic one the # of common queries to construct.
> Thinking restructuring this..
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

xtinec commented on superset/assets/src/query/buildQueries.ts:
```diff
@@ -0,0 +1,18 @@
+import { FormData, getGranularity } from './formData';
+
+// TODO: fill out the rest of the query object
```
It will go here as well. Planning on adding in a separate PR to limit the scope here. :)

kristw commented:
regarding 1) and 2). Before I write the comment I thought 2 but after written, I think let's go with 1 for now. We can switch to 2) easily. Just write another function to do the look up. And most likely this function will only be used by `query` so I was like why bother creating a separate one then.

Commit:
Further tweaking the generic buildQueryContext method and viz speicific buildQuery methodes per PR comments.

Commit:
git mv a renamed file.

kristw commented on superset/assets/src/query/buildDatasource.ts:
```diff
@@ -0,0 +1,35 @@
+}
+
+// Declaration merging with the interface above. No need to redeclare id and type.
+export class DatasourceKey {
```
One little tweak here. Let's `export default class DatasourceKey`, remove `buildDatasource` and make `buildQueryContext` call `new DatasourceKey(formData.datasource).toObject()` manually. 

I think that will make this file really clean (decoupled from `Formdata`).

kristw commented on superset/assets/src/query/FormData.ts:
```diff
@@ -0,0 +1,25 @@
+export default FormData;
+
+export function getGranularity(formData: FormData): string {
+  if ('granularity_sqla' in formData) {
```
nit. Could use ternary `? :` here

Commit:
addressing additional pr comments

kristw reviewed (approved):
Beautifully done!

williaster reviewed (approved):
👏 woot! this is a big one!
